### PR TITLE
ENH: Add the C99 remainder function, as `np.remainder_ieee`

### PR DESCRIPTION
**Pull request by eric-wieser**
Implementation stolen from cpython 3.7.

As suggested in a comment in #9702.

### CODE REVIEW

File: numpy/core/src/npymath/npy_math_internal.h.src [hunk 1]
```diff
@@ -672,6 +672,103 @@ npy_divmod@c@(@type@ a, @type@ b, @type@ *modulus)
...
+                                                  | 0, if n is even
+
+               Now m - 2.0 * fmod(...) gives the desired result: m
+               if n is even, -m if m is odd.
+
+               Note that all steps in fmod(0.5 * (absx - m), absy)
+               will be computed exactly, with no rounding error
+               introduced.
+            */
+            assert(m == c);
```
Comment (mhvk) [inline numpy/core/src/npymath/npy_math_internal.h.src]: left-over?

File: numpy/core/src/npymath/npy_math_internal.h.src [hunk 2]
```diff
@@ -672,6 +672,103 @@ npy_divmod@c@(@type@ a, @type@ b, @type@ *modulus)
...
+    if (npy_isnan(x)) {
+        return x;
+    }
+    if (npy_isnan(y)) {
+        return y;
+    }
+    if (npy_isfinite(x)) {
+        return NPY_NAN;
+    }
+    assert(npy_isfinite(y));
```
Comment (mhvk) [inline numpy/core/src/npymath/npy_math_internal.h.src]: Left-over from testing?

**Review by mhvk — DISMISSED** [review]
This looks good; just seems there are a few statements left in from testing.

I wish there was a way to change the meaning of the name, especially since we already have `np.mod` which is just the same thing as the current `np.remainder`. But the only way I can think of doing that is to add an argument to `np.remainder`, which could warn upon the "wrong" setting; this seems non-trivial for a ufunc.

**Comment by eric-wieser** [comment]
`assert`s are there in cpython too.

**Comment by eric-wieser** [comment]
Hang on, I think I might have muddled isfinire and isinfinite here

**Comment by eric-wieser** [comment]
Yep, I did. Is there a reason we have no `npy_isinfinite`? Or why we don't just use the python macros?

**Comment by rgommers** [comment]
I'm not a big fan of the name, nor of many almost identical but separate functions. Have you considered any alternatives, like ``np.remainder(..., method='ieee')``?

**Comment by charris** [comment]
@rgommers Maybe `pyremainder`? Adding a method keyword to a ufunc would be tricky. I'm not a big fan of the name either ...

**Comment by charris** [comment]
I suppose the function could be put somewhere else with a `method` keyword, but that is a more intrusive change.

**Comment by eric-wieser** [comment]
An argument for the name - it matches `java.lang.Math.IEEEremainder`

**Comment by mhvk** [comment]
@charris - I wondered about turning `remainder` into a front-end function as well, which either calls `mod` or the new remainder ufunc, but then we'd still need the `ufunc` anyway, so it seems that it is an independent discussion.

**Comment by rgommers** [comment]
I'd also like to see the argument for adding this in the first place. The differences with ``np.remainder`` seem small. Current examples:
```
In [1]: np.remainder([4, 7], [2, 3])
Out[1]: array([0, 1])

In [2]: np.remainder(np.arange(9), 4)
Out[2]: array([0, 1, 2, 3, 0, 1, 2, 3, 0])
```

Two differences; this implementation gives back floats rather than ints for int input, and the second one changes 3 to -1 in output. Unless I'm missing something major, that's a fairly marginal change.

**Comment by eric-wieser** [comment]
>this implementation gives back floats rather than ints for int input

This is an omission and not a feature.

> the second one changes 3 to -1 in output
> Unless I'm missing something major, that's a fairly marginal change.

This is as marginal a change as `round(1.5) == 2` vs `floor(1.5) == 1`. For each type of conversion to integer, there is a corresponding type of remainder.

So yes, there would be a good argument here for `np.remainder(..., rounding={'even', 'floor', 'ceil'})`, but in practice we'd want each of those to have their own ufunc for handling by `__array_ufunc__`

**Comment by rgommers** [comment]
> but then we'd still need the ufunc anyway, so it seems that it is an independent discussion.

Doesn't mean that it would need to be public though.

> n argument for the current name - it matches java.lang.Math.IEEEremainder

Hmm... we want to choose nice functional names, not match other languages. And of those others, Java is very low on the recognition scale so not very relevant.

**Comment by eric-wieser** [comment]
> Hmm... we want to choose nice functional names, not match other languages. And of those others, Java is very low on the recognition scale so not very relevant.

However, we also don't want to mismatch other languages, which is the problem `remainder` currently has. But yes, this is not a strong argument for `ieee_remainder`.

> Doesn't mean that it would need to be public though.

It needs to be visible to `__array_ufunc__`, which is almost public


I suppose we could move the functions to:

```
np.remainder(..., rounding='floor')
np.remainder.floor # old remainder
np.remainder.to_zero # mod
np.remainder.even   #this
```

**Comment by rgommers** [comment]
> This is as marginal a change as `round(1.5) == 2` vs `floor(1.5) == 1`.

Not quite I think. ``round`` and ``floor`` are much more common operations. If the majority opinion is to add this function I won't block it, but my opinion is that this is too marginal to add to the main numpy namespace.

**Comment by eric-wieser** [comment]
> my opinion is that this is too marginal to add to the main numpy namespace.

Does that opinion also stand against `np.remainder(..., rounding='even')` , or simply against the function having it's own name.

I suppose this is similar to #9696, in that python adding things to `math` is not a sufficient justification to us adding things to `np`

**Comment by rgommers** [comment]
> Does that opinion also stand against np.remainder(..., rounding='even') , or simply against the function having it's own name.

Adding as an extra method inside ``np.remainder`` would be much better, I'd be in favor of that. Reason: we get the same extra functionality, but not the downsides of namespace size growth.

**Comment by eric-wieser** [comment]
Unfortunately adding the extra method is a lot of work without breaking compatibility of `isinstance(np.remainder, np.ufunc)` and all the properties that entails.

Would you be happier with `np.remainder_ieee`, which is more discoverable by autocomplete?

**Comment by rgommers** [comment]
> Would you be happier with `np.remainder_ieee`, which is more discoverable by autocomplete?

Better, but doesn't make much of a difference to my opinion. If we're bikeshedding names, I don't think something like `ieee` belongs in a function name; names should be functional not named after entities.

**Comment by charris** [comment]
@rgommers I don't see the difficulty here. I think we need the function for completeness, both for Python compatibility and because is has greater precision on the near negative side of zero, which is where it is useful due to the distribution of floating point numbers. The main question is how to name it so that it does not collide with our long time `remainder` function and so that it carries some meaning for the user. I think `remainder_ieee` is about as good as we are going to get in that direction. If we later decide to add a pseudo ufunc with a method, we will still need an underlying ufunc, and I see no reason to hide it.

**Comment by charris** [comment]
Hmm, I suppose `__array_ufunc__` would still get called for the underlying computations, so perhaps a higher level implementation would not be so bad.

**Comment by mhvk** [comment]
@charris - I'd personally like turning `remainder` in a function that has a keyword that decides which underlying `ufunc` to call, and, as you noted, those ufuncs would be properly caught with `__array_ufunc__`. The only real annoyance is that, as @eric-wieser mentioned, this would break `isinstance(remainder, ufunc)`.

In any case, I think we should have the specific ufunc and might then as well expose it. `remainder_ieee` is a bit better. I guess `remainder_from_rounding` or `remainder_from_nearest` or `remainder_to_nearest` are all rather too long?

**Comment by charris** [comment]
Another option is to let folks roll their own, `a - around(a)`. However, there may be subtleties that such a simple implementation misses, for example, it is supposed to be independent of rounding mode.

**Comment by mhvk** [comment]
I do think we should have this as a `ufunc` exactly because of those subtleties. For this PR, I think the main question really is about the name!

**Comment by eric-wieser** [comment]
It sounds like `remainder_ieee` is preferred, so I'll update this. I think starting with `remainder` is probably more valuable than matching the java name.

**Comment by rgommers** [comment]
> @rgommers I don't see the difficulty here. 

It's simple math to me:
- people that have a use for this function: close to zero
- costs: more cognitive load for users, maintenance overhead for us

> I think we need the function for completeness, both for Python compatibility and because is has greater precision on the near negative side of zero, which is where it is useful due to the distribution of floating point numbers.

Completeness is in the eye of the beholder. This is nothing like rounding, where we clearly do need the two behaviors for rounding .5. `remainder` is not a heavily used function, and >99% of our users will not even know of this definition of it (I certainly didn't either). Of those that do, there's again only a small fraction that would need the last ulps of precision that you cannot get from the simple one-liner.

File: numpy/core/src/npymath/npy_math_internal.h.src (hunk 1 above)
Comment (charris) [reply numpy/core/src/npymath/npy_math_internal.h.src]: Doesn't hurt and might turn up potential problems. This comes from the Python implementation.

File: numpy/core/tests/test_umath.py [hunk 3]
```diff
@@ -1268,6 +1268,25 @@ def test_heaviside(self):
         assert_equal(h, expected1.astype(np.float32))
 
 
+class TestIeeeRemainder(object):
+    def test_remainder_ieee(self):
+        # Testing is basic because implementation is copied from cpython, and
```
Comment (charris) [inline numpy/core/tests/test_umath.py]: We need to test the different float types, no?

File: numpy/core/tests/test_umath.py [hunk 4]
```diff
@@ -1268,6 +1268,25 @@ def test_heaviside(self):
         assert_equal(h, expected1.astype(np.float32))
 
 
+class TestIeeeRemainder(object):
```
Comment (charris) [inline numpy/core/tests/test_umath.py]: Needs rename?

**Comment by charris** [comment]
I think we should support most C99 functions because if someone wants to generalize a C, or even a Python 3.7+ program,  to use ndarrays, then they may need the function. I do think this needs more tests.

**Comment by eric-wieser** [comment]
> I do think this needs more tests.

Ideally we'd copy the python3 tests there, but to do so we need to implement #9969.

**Comment by mattip** [comment]
@mhvk is there a way to remove your "review approval" from this PR? The idea seems still controversial

**Comment by eric-wieser** [comment]
Needs #13375 to port the tests too

**Comment by eric-wieser** [comment]
Now that C99 is a requirement, I'm inclined to remove the manual implementation, which also removes the need for detailed testing.

**Pull request closed without merging**